cpp/src/fil/infer.cu

```diff
@@ -248,20 +248,33 @@ struct tree_aggregator_t<NITEMS, GROVE_PER_CLASS_FEW_CLASSES> : finalize_block {
     __syncthreads();  // free up input row
     // load margin into shared memory
     auto per_thread = (vec<NITEMS, float>*)tmp_storage;
-    if (threadIdx.x >= num_classes) per_thread[threadIdx.x] = acc;
+    per_thread[threadIdx.x] = acc;
 
     __syncthreads();
     // reduce per-thread margin summand into per-class complete margin
     // (for each of the NITEMS rows)
-    // TODO(levsnv): use CUB/tree reduction when num_classes is small
-    if (threadIdx.x < num_classes) {
-      for (int c = threadIdx.x + num_classes; c < blockDim.x; c += num_classes)
-        acc += per_thread[c];
+    int t = threadIdx.x;
+    int sets = blockDim.x / num_classes;
+    for (; sets > 3; sets = sets / 4 + sets % 4) {
+      if (t < sets / 4 * num_classes) {
+        int add_start = t + sets % 4 * num_classes;
+        per_thread[t] = acc =
+          per_thread[t] + per_thread[add_start + sets / 4 * num_classes] +
+          per_thread[add_start + sets / 4 * 2 * num_classes] +
+          per_thread[add_start + sets / 4 * 3 * num_classes];
+      }
+      __syncthreads();
     }
-    __syncthreads();  // free up per_thread[] margin
-
-    write_best_class_in_block(to_vec(threadIdx.x, acc), num_classes, out,
-                              num_rows);
+    switch (sets) {
```

Comment:
Can this be handled within the main reduction loop without significant performance loss?

Reply:
if I changed the condition `sets > 3`, it would become an infinite loop for these cases. I do not see in any way how this can be merged in